### ISSUE TITLE
clcache statistics seem wrong

### ISSUE THREAD

**Issue by tadeu** [frerich/clcache#98]
I'm using `clcache` together with `CMake` and `ninja`, and noted that it is printing weird statistics.

Here is the workflow:

```
> mkdir build
> cd build
> set CC=clcache
> set CXX=clcache
> cmake -G Ninja ..
> clcache -C
> clcache -z
> ninja clean
> ninja
> clcache -s
```

THe `clcache -C` is just to ensure a clean cache in this verification process.
I noticed that in the end the cache is correctly populated with 55 objects, but the statistics only show 4 cache entries:

```
clcache statistics:
  current cache dir         : C:\Users\tadeu\clcache
  cache size                : 73,960,833 bytes
  maximum cache size        : 34,359,738,368 bytes
  cache entries             : 4
  cache hits                : 0
  cache misses
    total                      : 4
    evicted                    : 0
    header changed             : 0
    source changed             : 0
  passed to real compiler
    called for linking         : 0
    called for external debug  : 0
    called w/o source          : 0
    called w/ multiple sources : 0
    called w/ PCH              : 0
```

After a new `ninja clean` and `ninja` (which is a lot faster, due to `clcache` hits), the `clcache -s` statistics show:

```
clcache statistics:
  current cache dir         : C:\Users\tadeu\clcache
  cache size                : 73,960,833 bytes
  maximum cache size        : 34,359,738,368 bytes
  cache entries             : 4
  cache hits                : 5
  cache misses
    total                      : 4
    evicted                    : 0
    header changed             : 0
    source changed             : 0
  passed to real compiler
    called for linking         : 0
    called for external debug  : 0
    called w/o source          : 0
    called w/ multiple sources : 0
    called w/ PCH              : 0
```

**Comment by frerich** [comment]
This sounds like a duplicate of issue #70 -- it should be fixed by now. Can you try clcache 3.1.0 (which was released just today), does that work better?

**Comment by tadeu** [comment]
Ops, sorry about duplicating the issue, should we move the conversation to there?
This was already tested on `v3.1.0`.

**Comment by tadeu** [comment]
I'm just guessing here, but the parallel test is testing just compilation, but perhaps the problem is with parallel linkage too.

**Comment by tadeu** [comment]
Also, a difference here is that I'm using `CLCACHE_NODIRECT`

**Comment by frerich** [comment]
Interesting! No, no need to resurrect #70. Maybe the fixes (and the tests) which were done did not cover all cases.

* Can you try building with `ninja -j1`, does that still exhibit the issue? I wonder whether there's some other problem here which is unrelated to concurrency

* Is the project you're building available somewhere? I'd like to give it a try myself and check whether I can reproduce the issue.

**Comment by tadeu** [comment]
@frerich, building with `ninja -j1` the `cache size`, `cache entries` and `cache misses` are correct.

```
clcache statistics:                                 
  current cache dir         : C:\Users\tadeu\clcache
  cache size                : 146,695,108 bytes     
  maximum cache size        : 32,212,254,720 bytes  
  cache entries             : 55                    
  cache hits                : 0                     
  cache misses                                      
    total                      : 55                 
    evicted                    : 0                  
    header changed             : 0                  
    source changed             : 0                  
  passed to real compiler                           
    called for linking         : 0                  
    called for external debug  : 0                  
    called w/o source          : 0                  
    called w/ multiple sources : 0                  
    called w/ PCH              : 0                  
```

Sorry, it's not open source, I can't make the sources available, but I can help doing some more tests if you need, or provide some logs. The project consists of 55 object files, 7 shared libraries (DLL + import LIB), and one static library.

**Comment by frerich** [comment]
Alas, I currently only have access to a dual core machine -- where I can't seem to trigger this issue. I wonder whether it is more likely to happen if you have more CPU cores, given that artificially limiting the number of concurrent invocations of `clcache` via `ninja -j1` avoids the issue... I'll get some tea and go through the code again (depending on how exciting tonights football game is!), maybe I can spot something.

**Comment by frerich** [comment]
I also wonder how many CPU cores the virtual machines launched by the CI builds get. Maybe the parallel execution test case does not hit this issue because nothing ever gets executed in parallel.

**Comment by webmaster128** [comment]
> I also wonder how many CPU cores the virtual machines launched by the CI builds get.

[1](https://www.appveyor.com/docs/build-configuration#virtual-machine-configurations)

**Comment by tadeu** [comment]
But anyway, I've run the test locally, and it passed. I even added more compilation units that would take more time to compile, but it was still passing.

**Comment by tadeu** [comment]
Note that is seems to get worse the more parallel jobs are used. This is the result for `ninja -j2`:

```
clcache statistics:
  current cache dir         : C:\Users\tadeu\clcache
  cache size                : 85,260,019 bytes
  maximum cache size        : 32,212,254,720 bytes
  cache entries             : 37
  cache hits                : 0
  cache misses
    total                      : 37
    evicted                    : 0
    header changed             : 0
    source changed             : 0
  passed to real compiler
    called for linking         : 0
    called for external debug  : 0
    called w/o source          : 0
    called w/ multiple sources : 0
    called w/ PCH              : 0
```

**Comment by frerich** [comment]
It almost sounds like you get no locking at all!

I just pushed a new `debug/issue98` branch which is basically the current `master` branch plus some fairly noisy debug output related to the statistics reading/writing and cache locking. Can you please reproduce the problem using that branch and post the output? I hope to spot a problematic sequence of cache lock acquire/release calls, or maybe some missing calls to the code which saves the stats.

**Comment by tadeu** [comment]
Sorry people, major mistake here, after changing to the debug version and noting it wasn't printing anything (and certifying the output wasn't being suppressed), I noticed that the `rules.ninja` file (generated by CMake) had a _hardcoded_ PATH to `clcache.exe`, which was pointing to a an older `clcache.exe` in another location.

So, even if `clcache --version` was returning the newer version (the PATH was correct, as were `CC`and `CXX` environment variable), `ninja` was still using the older one :/.

I've checked with the newer version and it is indeed correct, so feel free to close this.

**Comment by frerich** [comment]
No worries, glad to hear that it had a plausible explanation after all!
